src/replica/replica_throttle.cpp

```diff
@@ -14,40 +14,49 @@
 namespace dsn {
 namespace replication {
 
-bool replica::throttle_request(throttling_controller &controller,
-                               message_ex *request,
-                               int32_t request_units)
+#define THROTTLE_REQUEST(op_type, throttling_type, request, request_units)                         \
+    do {                                                                                           \
+        int64_t delay_ms = 0;                                                                      \
+        auto type = _##op_type##_##throttling_type##_throttling_controller.control(                \
+            request->header->client.timeout_ms, request_units, delay_ms);                          \
+        if (type != throttling_controller::PASS) {                                                 \
+            if (type == throttling_controller::DELAY) {                                            \
+                tasking::enqueue(                                                                  \
+                    LPC_##op_type##_THROTTLING_DELAY,                                              \
+                    &_tracker,                                                                     \
+                    [ this, req = message_ptr(request) ]() { on_client_##op_type(req, true); },    \
+                    get_gpid().thread_hash(),                                                      \
+                    std::chrono::milliseconds(delay_ms));                                          \
+                _counter_recent_##op_type##_throttling_delay_count->increment();                   \
+            } else { /** type == throttling_controller::REJECT **/                                 \
+                if (delay_ms > 0) {                                                                \
+                    tasking::enqueue(LPC_##op_type##_THROTTLING_DELAY,                             \
+                                     &_tracker,                                                    \
+                                     [ this, req = message_ptr(request) ]() {                      \
+                                         response_client_##op_type(req, ERR_BUSY);                 \
+                                     },                                                            \
+                                     get_gpid().thread_hash(),                                     \
+                                     std::chrono::milliseconds(delay_ms));                         \
+                } else {                                                                           \
+                    response_client_##op_type(request, ERR_BUSY);                                  \
```

Comment:
I think you can update throttling error code in this pull request.

Reply:
I think we can update it someday later. Modify at the same time as java client